File: .github/scripts/audit.py
#!/usr/bin/env python3
"""Automated code audit for GRG Mobile repository."""

import os
import subprocess
import urllib.request
import urllib.parse
from pathlib import Path


def get_changed_files():
    event = os.environ.get('GITHUB_EVENT_NAME', 'push')

    if event == 'pull_request':
        base = os.environ.get('PR_BASE_SHA', '')
        head = os.environ.get('PR_HEAD_SHA', '')
    else:
        base = os.environ.get('GITHUB_BEFORE', '')
        head = os.environ.get('GITHUB_SHA', '')

    if not base or base == '0000000000000000000000000000000000000000':
        result = subprocess.run(
            ['git', 'show', '--name-only', '--format=', 'HEAD'],
            capture_output=True, text=True
        )
    else:
        result = subprocess.run(
            ['git', 'diff', '--name-only', base, head],
            capture_output=True, text=True
        )

    return [f for f in result.stdout.strip().split('\n') if f]


def read_file(path):
    try:
        return Path(path).read_text(errors='ignore') if Path(path).exists() else ''
    except Exception:
        return ''


def analyze_backend(files):
    info, issues = [], []
    backend_files = [f for f in files if f.startswith('backend/')]
    if not backend_files:
        return None, []

    migration_files = [f for f in backend_files if 'migration' in f.lower()]
    entity_files = [f for f in backend_files if f.endswith('.entity.ts')]
    module_files = [f for f in backend_files if f.endswith('.module.ts') and 'app.module' not in f]
    app_module_changed = 'backend/src/app.module.ts' in backend_files
    schema_files = [f for f in backend_files if 'schema' in f.lower() and f.endswith('.sql')]

    # New modules without app.module.ts update
    if module_files and not app_module_changed:
        names = ', '.join(Path(f).name for f in module_files)
        issues.append(f"⚠️ Новые модули ({names}) — app.module.ts не изменён, возможно модули не зарегистрированы")

    # Entity changes without migrations
    if entity_files and not migration_files:
        names = ', '.join(Path(f).stem for f in entity_files)
        issues.append(f"⚠️ Изменены сущности TypeORM ({names}), но файлы миграций не найдены")

    # Check credential handling in entity files
    for f in entity_files:
        content = read_file(f)
        if ('password' in content or 'username' in content) and 'credentials' not in f.lower():
            if 'CredentialsService' not in content and 'encrypt' not in content.lower():
                issues.append(f"⚠️ {Path(f).name}: поля credentials без шифрования через CredentialsService")

    # Check SQL migration files for up/down pattern
    for f in migration_files:
        content = read_file(f)
        if content and 'down' not in content.lower() and f.endswith('.ts'):
            issues.append(f"⚠️ {Path(f).name}: миграция без метода down() — откат невозможен")

    # Access control check in controllers
    ctrl_files = [f for f in backend_files if f.endswith('.controller.ts')]
    for f in ctrl_files:
        content = read_file(f)
        if '@Get' in content or '@Post' in content or '@Put' in content or '@Delete' in content:
            if '@UseGuards' not in content and 'JwtAuthGuard' not in content:
                issues.append(f"⚠️ {Path(f).name}: эндпоинты без @UseGuards(JwtAuthGuard) — проверьте защиту маршрутов")

    info.append(f"📁 Backend файлов: {len(backend_files)}")
    if entity_files:
        info.append(f"📋 Сущности: {', '.join(Path(f).stem for f in entity_files)}")
    if migration_files:
        info.append(f"✅ Миграции: {', '.join(Path(f).name for f in migration_files)}")
    if schema_files:
        info.append(f"🗄️ SQL-схемы: {', '.join(Path(f).name for f in schema_files)}")
    if ctrl_files:
        info.append(f"🔌 Контроллеры: {', '.join(Path(f).stem for f in ctrl_files)}")

    return info, issues


def analyze_flutter(files):
    info, issues = [], []
    flutter_files = [f for f in files if f.startswith('lib/') or f == 'pubspec.yaml']
    if not flutter_files:
        return None, []

    screen_files = [f for f in flutter_files if '/screens/' in f]
    widget_files = [f for f in flutter_files if '/widgets/' in f]
    model_files = [f for f in flutter_files if '/models/' in f]
    service_files = [f for f in flutter_files if '/services/' in f]
    api_files = [f for f in flutter_files if '/api/' in f]

    # Files in lib/ root (outside expected subdirs)
    allowed_root = {'main.dart', 'firebase_options.dart'}
    for f in flutter_files:
        if f.startswith('lib/') and f.endswith('.dart') and f.count('/') == 1:
            if Path(f).name not in allowed_root:
                issues.append(f"⚠️ {f}: файл в корне lib/ — рекомендуется screens/, widgets/, models/ или services/")

    # Check null safety and force-unwrap usage
    for f in flutter_files:
        if not f.endswith('.dart'):
            continue
        content = read_file(f)
        if not content:
            continue
        force_unwraps = content.count('!.')
        if force_unwraps > 8:
            issues.append(f"⚠️ {Path(f).name}: {force_unwraps} использований '!.' — проверьте null safety")
        # Check for hardcoded IPs or tokens
        if 'http://' in content and ('192.168.' in content or '10.0.' in content):
            issues.append(f"⚠️ {Path(f).name}: возможный хардкод IP-адреса в коде")

    # pubspec.yaml changes
    if 'pubspec.yaml' in flutter_files:
        content = read_file('pubspec.yaml')
        info.append("📦 pubspec.yaml изменён — проверьте совместимость зависимостей")

    info.append(f"📱 Flutter файлов: {len(flutter_files)}")
    if screen_files:
        info.append(f"🖥️ Экраны: {', '.join(Path(f).stem for f in screen_files)}")
    if widget_files:
        info.append(f"🧩 Виджеты: {', '.join(Path(f).stem for f in widget_files)}")
    if model_files:
        info.append(f"📊 Модели: {', '.join(Path(f).stem for f in model_files)}")
    if service_files:
        info.append(f"⚙️ Сервисы: {', '.join(Path(f).stem for f in service_files)}")
    if api_files:
        info.append(f"🌐 API клиенты: {', '.join(Path(f).stem for f in api_files)}")

    return info, issues


def analyze_go2rtc(files):
    info, issues = [], []
    streaming_files = []

    keywords = ('go2rtc', 'rtsp', 'live-url', 'liveurl', 'live_url', 'livestream')

    for f in files:
        content = read_file(f)
        if not content:
            continue
        if any(kw in content.lower() for kw in keywords):
            streaming_files.append(f)

    if not streaming_files:
        return None, []

    for f in streaming_files:
        content = read_file(f)
        if not content:
            continue

        if f.endswith('.ts') or f.endswith('.js'):
            # Missing try/catch
            if 'try' not in content or 'catch' not in content:
                issues.append(f"⚠️ {Path(f).name}: go2rtc/RTSP вызовы без try/catch")
            # Missing HTTP error propagation
            if 'throw' not in content and 'HttpException' not in content:
                issues.append(f"⚠️ {Path(f).name}: нет выброса HTTP-исключений при ошибках стриминга")
            # Timeout check
            if 'timeout' not in content.lower():
                issues.append(f"⚠️ {Path(f).name}: нет таймаута для go2rtc/RTSP запросов")

    info.append(f"🎥 Файлы со стримингом: {', '.join(Path(f).name for f in streaming_files)}")
    return info, issues


def send_telegram(text, token, chat_id):
    url = f"https://api.telegram.org/bot{token}/sendMessage"
    data = urllib.parse.urlencode({
        'chat_id': chat_id,
        'text': text,
        'parse_mode': 'HTML',
        'disable_web_page_preview': 'true',
    }).encode('utf-8')
    try:
        req = urllib.request.Request(url, data=data, method='POST')
        with urllib.request.urlopen(req, timeout=30) as resp:
            return resp.status == 200
    except Exception as e:
        print(f"Telegram error: {e}")
        return False


def main():
    token = os.environ.get('TELEGRAM_BOT_TOKEN', '')
    chat_id = os.environ.get('TELEGRAM_CHAT_ID', '')
    event = os.environ.get('GITHUB_EVENT_NAME', 'push')
    ref = os.environ.get('GITHUB_REF', '')
    sha = os.environ.get('GITHUB_SHA', '')[:7]
    actor = os.environ.get('ACTOR', 'unknown')
    repo = os.environ.get('REPO', '')
    pr_title = os.environ.get('PR_TITLE', '')
    pr_number = os.environ.get('PR_NUMBER', '')

    if not token or not chat_id:
        print("TELEGRAM_BOT_TOKEN or TELEGRAM_CHAT_ID not set — skipping report")
        return

    changed = get_changed_files()
    if not changed:
        print("No changed files")
        return

    backend_info, backend_issues = analyze_backend(changed)
    flutter_info, flutter_issues = analyze_flutter(changed)
    go2rtc_info, go2rtc_issues = analyze_go2rtc(changed)

    all_issues = backend_issues + flutter_issues + go2rtc_issues
    branch = ref.replace('refs/heads/', '').replace('refs/pull/', 'PR/')

    lines = ['<b>🔍 Аудит GRG Mobile</b>', '']

    if event == 'pull_request':
        lines += [f'<b>Событие:</b> Pull Request #{pr_number}', f'<b>Заголовок:</b> {pr_title}']
    else:
        lines += [f'<b>Событие:</b> Push', f'<b>Ветка:</b> {branch}']

    lines += [
        f'<b>Коммит:</b> <code>{sha}</code>',
        f'<b>Автор:</b> {actor}',
        f'<b>Репозиторий:</b> {repo}',
        f'<b>Изменено файлов:</b> {len(changed)}',
        '',
    ]

    if backend_info:
        lines.append('<b>🖥️ Backend (NestJS / PostgreSQL):</b>')
        lines += [f'  {i}' for i in backend_info]
        lines.append('')

    if flutter_info:
        lines.append('<b>📱 Flutter UI:</b>')
        lines += [f'  {i}' for i in flutter_info]
        lines.append('')

    if go2rtc_info:
        lines.append('<b>🎥 Видеостриминг (go2rtc/RTSP):</b>')
        lines += [f'  {i}' for i in go2rtc_info]
        lines.append('')

    if all_issues:
        lines.append(f'<b>🚨 Найдено проблем: {len(all_issues)}</b>')
        lines += [f'  {i}' for i in all_issues]
        lines.append('')
        lines.append('<b>Соответствие архитектуре:</b> ⚠️ Требует внимания')
    else:
        lines.append('<b>✅ Критических проблем не обнаружено</b>')
        lines.append('<b>Соответствие архитектуре:</b> ✅ Соответствует')

    report = '\n'.join(lines)
    if len(report) > 4000:
        report = report[:3970] + '\n\n<i>... отчёт обрезан (превышен лимит)</i>'

    ok = send_telegram(report, token, chat_id)
    print("Report sent." if ok else "Failed to send report.")


if __name__ == '__main__':
    main()

File: .github/scripts/audit.js
#!/usr/bin/env node
/**
 * GRG Mobile — Automated Audit Script
 * Triggered by GitHub Actions on push / pull_request.
 * Analyses changed files and sends a Telegram report.
 */

const { execSync } = require('child_process');
const https = require('https');

// ── Environment ─────────────────────────────────────────────────────────────
const TELEGRAM_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const TELEGRAM_CHAT  = process.env.TELEGRAM_CHAT_ID;
const EVENT_NAME     = process.env.GITHUB_EVENT_NAME  || 'unknown';
const REPO           = process.env.GITHUB_REPOSITORY  || 'unknown/repo';
const REF            = process.env.GITHUB_REF_NAME    || '';
const SHA            = (process.env.GITHUB_SHA        || '').slice(0, 8);
const ACTOR          = process.env.GITHUB_ACTOR       || 'unknown';
const PR_TITLE       = process.env.PR_TITLE           || '';
const PR_NUMBER      = process.env.PR_NUMBER          || '';
const BASE_SHA       = process.env.BASE_SHA           || '';
const HEAD_SHA       = process.env.HEAD_SHA           || '';

if (!TELEGRAM_TOKEN || !TELEGRAM_CHAT) {
  console.error('Missing TELEGRAM_BOT_TOKEN or TELEGRAM_CHAT_ID');
  process.exit(1);
}

// ── Helpers ──────────────────────────────────────────────────────────────────
function run(cmd) {
  try {
    return execSync(cmd, { encoding: 'utf8', stdio: ['pipe', 'pipe', 'pipe'] }).trim();
  } catch {
    return '';
  }
}

function getChangedFiles() {
  let files = '';
  if (BASE_SHA && HEAD_SHA) {
    files = run(`git diff --name-only ${BASE_SHA} ${HEAD_SHA}`);
  } else {
    files = run('git diff --name-only HEAD~1 HEAD');
  }
  return files ? files.split('\n').filter(Boolean) : [];
}

function fileContains(path, pattern) {
  try {
    const content = run(`git show HEAD:${path} 2>/dev/null || cat ${path} 2>/dev/null`);
    return pattern.test(content);
  } catch {
    return false;
  }
}

function getDiff(path) {
  if (BASE_SHA && HEAD_SHA) {
    return run(`git diff ${BASE_SHA} ${HEAD_SHA} -- "${path}"`);
  }
  return run(`git diff HEAD~1 HEAD -- "${path}"`);
}

// ── Categorise files ─────────────────────────────────────────────────────────
function categorise(files) {
  return {
    backend:    files.filter(f => f.startsWith('backend/')),
    flutter:    files.filter(f => f.startsWith('lib/') || f.endsWith('.dart')),
    migrations: files.filter(f => /migration|\.sql$/i.test(f)),
    entities:   files.filter(f => /\.entity\.ts$/.test(f)),
    schemas:    files.filter(f => /schema|dto/i.test(f) && f.endsWith('.ts')),
    streaming:  files.filter(f => /go2rtc|stream|rtsp|live/i.test(f)),
    modules:    files.filter(f => /\.module\.ts$/.test(f)),
    controllers: files.filter(f => /\.controller\.ts$/.test(f)),
    services:   files.filter(f => /\.service\.ts$/.test(f)),
    workflows:  files.filter(f => f.startsWith('.github/')),
    configs:    files.filter(f => /\.(env|json|yaml|yml)$/.test(f) && !f.startsWith('.github/')),
  };
}

// ── Audit: Backend ────────────────────────────────────────────────────────────
function auditBackend(cat) {
  const issues = [];
  const ok     = [];

  // Migrations
  if (cat.migrations.length > 0) {
    cat.migrations.forEach(f => {
      const diff = getDiff(f);
      if (/DROP TABLE|DROP COLUMN|ALTER COLUMN/i.test(diff)) {
        issues.push(`⚠️ Деструктивная миграция в \`${f}\` (DROP/ALTER — проверь обратную совместимость)`);
      } else {
        ok.push(`✅ Миграция \`${f}\` — только аддитивные изменения`);
      }
    });
  }

  // Entities
  if (cat.entities.length > 0) {
    cat.entities.forEach(f => {
      const diff = getDiff(f);
      const hasColumnChange = /@Column|@PrimaryGeneratedColumn|@ManyToOne|@OneToMany|@ManyToMany/.test(diff);
      const hasMigration    = cat.migrations.length > 0;
      if (hasColumnChange && !hasMigration) {
        issues.push(`⚠️ Изменена схема сущности \`${f}\`, но миграция не найдена (только synchronize:true спасёт в dev)`);
      } else if (hasColumnChange) {
        ok.push(`✅ Сущность \`${f}\` изменена — сопровождается миграцией`);
      }
    });
  }

  // Module imports check
  if (cat.services.length > 0) {
    cat.services.forEach(f => {
      const diff = getDiff(f);
      // New injectable service added — check module wiring
      if (/\+.*@Injectable/.test(diff)) {
        const hasModuleChange = cat.modules.length > 0;
        if (!hasModuleChange) {
          issues.push(`⚠️ Новый @Injectable в \`${f}\`, но .module.ts не изменён — возможна UnknownDependenciesException`);
        } else {
          ok.push(`✅ Новый сервис \`${f}\` — модуль обновлён`);
        }
      }
    });
  }

  // Controllers — auth guard check
  if (cat.controllers.length > 0) {
    cat.controllers.forEach(f => {
      const diff = getDiff(f);
      const newRoutes = (diff.match(/^\+.*@(Get|Post|Put|Patch|Delete)\(/gm) || []).length;
      if (newRoutes > 0) {
        const hasGuard = /@UseGuards|@Roles|JwtAuthGuard/.test(diff);
        if (!hasGuard) {
          issues.push(`⚠️ ${newRoutes} новых маршрутов в \`${f}\` без явных guard-декораторов — проверь авторизацию`);
        } else {
          ok.push(`✅ Новые маршруты в \`${f}\` защищены guard-ами`);
        }
      }
    });
  }

  // Access control pattern
  cat.backend.filter(f => f.endsWith('.service.ts')).forEach(f => {
    const diff = getDiff(f);
    if (/\+.*find(All|One|Many)/.test(diff) && !/findAllForUser|findByIdForUser|accessService/.test(diff)) {
      issues.push(`⚠️ Новый find-запрос в \`${f}\` — убедись, что используется фильтрация по пользователю (RBAC)`);
    }
  });

  return { issues, ok };
}

// ── Audit: Flutter ────────────────────────────────────────────────────────────
function auditFlutter(cat) {
  const issues = [];
  const ok     = [];

  cat.flutter.forEach(f => {
    const diff = getDiff(f);

    // API calls without error handling
    if (/BackendClient\.|http\.get|http\.post|dio\./.test(diff)) {
      if (!/(try\s*\{|\.catchError|onError)/.test(diff)) {
        issues.push(`⚠️ HTTP-вызов в \`${f}\` без try/catch или catchError`);
      } else {
        ok.push(`✅ HTTP-вызов в \`${f}\` обёрнут в обработчик ошибок`);
      }
    }

    // Navigator without mounted check
    if (/Navigator\.(push|pop|pushReplacement)/.test(diff) && !/mounted/.test(diff)) {
      issues.push(`⚠️ Navigator вызывается в \`${f}\` без проверки mounted (утечка контекста)`);
    }

    // Hardcoded API URLs
    if (/["']https?:\/\/[^'"]+["']/.test(diff) && !/ApiConfig|localhost/.test(diff)) {
      issues.push(`⚠️ Захардкоженный URL в \`${f}\` — используй ApiConfig`);
    }

    // setState in async without mounted
    if (/setState\(/.test(diff) && /async/.test(diff) && !/mounted/.test(diff)) {
      issues.push(`⚠️ setState после await в \`${f}\` без проверки mounted`);
    }
  });

  // Check pubspec for new dependencies
  if (cat.configs.some(f => f === 'pubspec.yaml')) {
    ok.push(`ℹ️ pubspec.yaml изменён — запусти flutter pub get перед тестированием`);
  }

  return { issues, ok };
}

// ── Audit: Streaming / go2rtc ─────────────────────────────────────────────────
function auditStreaming(cat) {
  const issues = [];
  const ok     = [];

  cat.streaming.forEach(f => {
    const diff = getDiff(f);

    // WebSocket / stream error handling
    if (/WebSocket|ws\.connect|go2rtc/.test(diff)) {
      if (!/(onError|on\('error'|try\s*\{|catch)/.test(diff)) {
        issues.push(`🔴 WebSocket/stream в \`${f}\` — отсутствует обработка ошибок подключения`);
      } else {
        ok.push(`✅ WebSocket/stream в \`${f}\` имеет обработчик ошибок`);
      }
    }

    // Credential exposure in URLs
    if (/rtsp:\/\//.test(diff) && /password|passwd|pwd/.test(diff)) {
      issues.push(`🔴 RTSP URL в \`${f}\` содержит учётные данные в открытом виде`);
    }

    // go2rtc endpoint — timeout check
    if (/go2rtc|\/api\/streams/.test(diff) && !/timeout|TimeoutException/.test(diff)) {
      issues.push(`⚠️ go2rtc эндпоинт в \`${f}\` без таймаута запроса`);
    }

    // Reconnection logic
    if (/disconnect|close\(\)/.test(diff) && !/reconnect|retry/.test(diff)) {
      ok.push(`ℹ️ \`${f}\` — закрытие стрима без логики реконнекта (приемлемо для ручного управления)`);
    }
  });

  return { issues, ok };
}

// ── Security quick-scan ───────────────────────────────────────────────────────
function securityScan(files) {
  const issues = [];

  files.forEach(f => {
    const diff = getDiff(f);

    // Secrets in code
    if (/password\s*=\s*["'][^"']{4,}["']|secret\s*=\s*["'][^"']{4,}["']/i.test(diff) &&
        !/(process\.env|\.env|secrets\.)/.test(diff)) {
      issues.push(`🔴 Секрет или пароль в коде \`${f}\` — используй process.env / GitHub Secrets`);
    }

    // SQL injection risk
    if (/query\(.*\$\{/.test(diff) || /createQueryBuilder.*where.*\$\{/.test(diff)) {
      issues.push(`🔴 Возможная SQL-инъекция в \`${f}\` — используй параметризованные запросы`);
    }

    // eval / exec with user input
    if (/eval\(|exec\(.*req\.|execSync\(.*req\./.test(diff)) {
      issues.push(`🔴 Небезопасное использование eval/exec с пользовательскими данными в \`${f}\``);
    }
  });

  return issues;
}

// ── Architecture conformance ──────────────────────────────────────────────────
function archCheck(cat, allFiles) {
  const notes = [];

  // app.module.ts modified without module file change could mean manual wiring
  if (allFiles.includes('backend/src/app.module.ts') && cat.modules.length === 0) {
    notes.push(`ℹ️ app.module.ts изменён напрямую — предпочтительно выносить в отдельный *.module.ts`);
  }

  // New vendor file without integration in devices/control service
  const newVendorFiles = cat.backend.filter(f => f.includes('/vendors/'));
  if (newVendorFiles.length > 0) {
    const controlOrDeviceTouched = allFiles.some(f => f.includes('/control/') || f.includes('/devices/'));
    if (!controlOrDeviceTouched) {
      notes.push(`ℹ️ Новые файлы вендора (${newVendorFiles[0]}), но control/devices не изменены — vendor подключён к роутингу?`);
    }
  }

  // Direct credential usage check
  cat.backend.forEach(f => {
    const diff = getDiff(f);
    if (/username.*password|password.*username/.test(diff) && !/CredentialsService|decrypt/.test(diff)) {
      notes.push(`⚠️ \`${f}\` работает с учётными данными напрямую — используй CredentialsService`);
    }
  });

  return notes;
}

// ── Build report ──────────────────────────────────────────────────────────────
function buildReport(allFiles) {
  const cat        = categorise(allFiles);
  const backend    = auditBackend(cat);
  const flutter    = auditFlutter(cat);
  const streaming  = auditStreaming(cat);
  const security   = securityScan(allFiles);
  const arch       = archCheck(cat, allFiles);

  const allIssues = [
    ...security,
    ...backend.issues,
    ...flutter.issues,
    ...streaming.issues,
    ...arch,
  ];
  const allOk = [...backend.ok, ...flutter.ok, ...streaming.ok];

  const criticalCount = allIssues.filter(i => i.startsWith('🔴')).length;
  const warnCount     = allIssues.filter(i => i.startsWith('⚠️')).length;

  const statusEmoji = criticalCount > 0 ? '🚨' : warnCount > 0 ? '⚠️' : '✅';
  const statusText  = criticalCount > 0
    ? `КРИТИЧЕСКИЕ ПРОБЛЕМЫ (${criticalCount})`
    : warnCount > 0
    ? `Предупреждения (${warnCount})`
    : 'Всё чисто';

  const header = EVENT_NAME === 'pull_request'
    ? `${statusEmoji} *Аудит PR #${PR_NUMBER}* — ${escapeMarkdown(PR_TITLE)}`
    : `${statusEmoji} *Аудит push* в \`${escapeMarkdown(REF)}\``;

  const lines = [
    header,
    `📦 Репо: \`${REPO}\`  |  Автор: @${ACTOR}  |  SHA: \`${SHA}\``,
    `🏷 Статус: *${statusText}*`,
    '',
    `📂 *Изменено файлов:* ${allFiles.length}`,
  ];

  if (cat.backend.length)    lines.push(`  • Backend NestJS: ${cat.backend.length}`);
  if (cat.flutter.length)    lines.push(`  • Flutter: ${cat.flutter.length}`);
  if (cat.migrations.length) lines.push(`  • Миграции SQL: ${cat.migrations.length}`);
  if (cat.entities.length)   lines.push(`  • Сущности: ${cat.entities.length}`);
  if (cat.streaming.length)  lines.push(`  • Стриминг/go2rtc: ${cat.streaming.length}`);

  if (allIssues.length > 0) {
    lines.push('', '─────────────────────────');
    lines.push('🔍 *Найденные проблемы:*');
    allIssues.slice(0, 15).forEach(i => lines.push(i));
    if (allIssues.length > 15) lines.push(`... и ещё ${allIssues.length - 15} предупреждений`);
  }

  if (allOk.length > 0) {
    lines.push('', '─────────────────────────');
    lines.push('🟢 *Проверки пройдены:*');
    allOk.slice(0, 8).forEach(i => lines.push(i));
  }

  if (allIssues.length === 0 && allOk.length === 0) {
    lines.push('', 'ℹ️ Специфичных паттернов для аудита не обнаружено.');
  }

  lines.push('', `_Аудит выполнен автоматически · GRG Mobile CI_`);

  return lines.join('\n');
}

function escapeMarkdown(str) {
  return (str || '').replace(/[_*[\]()~`>#+\-=|{}.!]/g, '\\$&');
}

// ── Send Telegram message ─────────────────────────────────────────────────────
function sendTelegram(text) {
  return new Promise((resolve, reject) => {
    const body = JSON.stringify({
      chat_id:    TELEGRAM_CHAT,
      text,
      parse_mode: 'Markdown',
      disable_web_page_preview: true,
    });

    const options = {
      hostname: 'api.telegram.org',
      path:     `/bot${TELEGRAM_TOKEN}/sendMessage`,
      method:   'POST',
      headers:  {
        'Content-Type':   'application/json',
        'Content-Length': Buffer.byteLength(body),
      },
    };

    const req = https.request(options, res => {
      let data = '';
      res.on('data', chunk => (data += chunk));
      res.on('end', () => {
        const parsed = JSON.parse(data);
        if (!parsed.ok) {
          reject(new Error(`Telegram API error: ${JSON.stringify(parsed)}`));
        } else {
          resolve(parsed);
        }
      });
    });

    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
  console.log('🔍 GRG Mobile Audit starting…');

  const changedFiles = getChangedFiles();
  console.log(`Changed files (${changedFiles.length}):`, changedFiles);

  if (changedFiles.length === 0) {
    console.log('No changed files detected — skipping audit.');
    return;
  }

  const report = buildReport(changedFiles);
  console.log('\n── REPORT ──\n', report);

  await sendTelegram(report);
  console.log('✅ Report sent to Telegram.');
}

main().catch(err => {
  console.error('Audit failed:', err.message);
  process.exit(1);
});
